Python/specialize.c

```diff
@@ -2663,101 +2681,94 @@ _Py_Specialize_Send(_PyStackRef receiver_st, _Py_CODEUNIT *instr)
+    }
+    if (PyDict_CheckExact(value)) {
+        return SPEC_FAIL_TO_BOOL_DICT;
+    }
+    if (PyFloat_CheckExact(value)) {
+        return SPEC_FAIL_TO_BOOL_FLOAT;
+    }
+    if (PyMemoryView_Check(value)) {
+        return SPEC_FAIL_TO_BOOL_MEMORY_VIEW;
+    }
+    if (PyAnySet_CheckExact(value)) {
+        return SPEC_FAIL_TO_BOOL_SET;
+    }
+    if (PyTuple_CheckExact(value)) {
+        return SPEC_FAIL_TO_BOOL_TUPLE;
+    }
+    return SPEC_FAIL_OTHER;
+}
+
 void
 _Py_Specialize_ToBool(_PyStackRef value_o, _Py_CODEUNIT *instr)
 {
-    assert(ENABLE_SPECIALIZATION);
+    assert(ENABLE_SPECIALIZATION_FT);
     assert(_PyOpcode_Caches[TO_BOOL] == INLINE_CACHE_ENTRIES_TO_BOOL);
     _PyToBoolCache *cache = (_PyToBoolCache *)(instr + 1);
     PyObject *value = PyStackRef_AsPyObjectBorrow(value_o);
     if (PyBool_Check(value)) {
-        instr->op.code = TO_BOOL_BOOL;
-        goto success;
+        specialize(instr, TO_BOOL_BOOL);
+        return;
     }
     if (PyLong_CheckExact(value)) {
-        instr->op.code = TO_BOOL_INT;
-        goto success;
+        specialize(instr, TO_BOOL_INT);
+        return;
     }
     if (PyList_CheckExact(value)) {
-        instr->op.code = TO_BOOL_LIST;
-        goto success;
+        specialize(instr, TO_BOOL_LIST);
+        return;
     }
     if (Py_IsNone(value)) {
-        instr->op.code = TO_BOOL_NONE;
-        goto success;
+        specialize(instr, TO_BOOL_NONE);
+        return;
     }
     if (PyUnicode_CheckExact(value)) {
-        instr->op.code = TO_BOOL_STR;
-        goto success;
+        specialize(instr, TO_BOOL_STR);
+        return;
     }
     if (PyType_HasFeature(Py_TYPE(value), Py_TPFLAGS_HEAPTYPE)) {
         PyNumberMethods *nb = Py_TYPE(value)->tp_as_number;
         if (nb && nb->nb_bool) {
-            SPECIALIZATION_FAIL(TO_BOOL, SPEC_FAIL_TO_BOOL_NUMBER);
-            goto failure;
+            unspecialize(instr, SPEC_FAIL_TO_BOOL_NUMBER);
+            return;
         }
         PyMappingMethods *mp = Py_TYPE(value)->tp_as_mapping;
         if (mp && mp->mp_length) {
-            SPECIALIZATION_FAIL(TO_BOOL, SPEC_FAIL_TO_BOOL_MAPPING);
-            goto failure;
+            unspecialize(instr, SPEC_FAIL_TO_BOOL_MAPPING);
+            return;
         }
         PySequenceMethods *sq = Py_TYPE(value)->tp_as_sequence;
         if (sq && sq->sq_length) {
-            SPECIALIZATION_FAIL(TO_BOOL, SPEC_FAIL_TO_BOOL_SEQUENCE);
-            goto failure;
+            unspecialize(instr, SPEC_FAIL_TO_BOOL_SEQUENCE);
+            return;
         }
         if (!PyUnstable_Type_AssignVersionTag(Py_TYPE(value))) {
-            SPECIALIZATION_FAIL(TO_BOOL, SPEC_FAIL_OUT_OF_VERSIONS);
-            goto failure;
+            unspecialize(instr, SPEC_FAIL_OUT_OF_VERSIONS);
+            return;
         }
-        uint32_t version = type_get_version(Py_TYPE(value), TO_BOOL);
+        uint32_t version = FT_ATOMIC_LOAD_UINT32_RELAXED(Py_TYPE(value)->tp_version_tag);
```

Comment:
This block of code needs to happen atomically with respect to other threads mutating the type. In free-threaded builds that requires holding the type lock. Structuring the code might be a little tricky, because I don't think we want to expose the type lock and we probably also don't want to move this logic into `typeobject.c`. One possible arrangement might be something like:

```c
// Returns 0 on success or caller-specific error on failure.
typedef uint16_t (*_py_validate_type)(PyTypeObject *);

// Returns 0 on success, -1 if no type version could be assigned, or the error returned by validate
int
_PyType_Validate(PyTypeObject *ty, _py_validate_type validate, unsigned int *tp_version)
{
    int err;
    BEGIN_TYPE_LOCK();
    err = validate(ty);
    if (!err) {
        if(assign_version_tag(_PyInterpreterState_GET(), ty))
            *tp_version = ty->tp_version_tag;
        }
        else {
            err = -1;
        }
    }
    END_TYPE_LOCK();
    return err;
}

// The specialization code 

static uint16_t
check_type_always_true(PyTypeObject *ty)
{
    PyNumberMethods *nb = ty->tp_as_number;
    if (nb && nb->nb_bool) {
        return SPEC_FAIL_TO_BOOL_NUMBER;
    }
    PyMappingMethods *mp = Py_TYPE(value)->tp_as_mapping;
    if (mp && mp->mp_length) {
        return SPEC_FAIL_TO_BOOL_MAPPING;
    }
    PySequenceMethods *sq = Py_TYPE(value)->tp_as_sequence;
    if (sq && sq->sq_length) {
      return SPEC_FAIL_TO_BOOL_SEQUENCE;
    }
    return 0;
}


if (PyType_HasFeature(Py_TYPE(value), Py_TPFLAGS_HEAPTYPE)) {
    unsigned int version = 0;
    int err = _PyType_Validate(Py_TYPE(value), check_type_always_true, &version);
    if (err < 0) {
        unspecialize(instr, SPEC_FAIL_OUT_OF_VERSIONS);
    }
    elif (err > 0) {
        unspecialize(instr, err);
    }
    else {
        write_u32(cache->version, version);
        specialize(instr, TO_BOOL_ALWAYS_TRUE);
        assert(version);
    }
    return;
}
```

Reply:
Brilliant approach. Thank you for letting me know where I should care :)
I've addressed your code review!